### DISCUSSION TITLE
Dicom files are not opening in weasis when click on view button of studies in dcm4chee ui

### DISCUSSION THREAD

**Discussion by abdullah951** [dcm4che/dcm4chee-arc-light#4748] [General]
![Screenshot 2024-10-24 203259](https://github.com/user-attachments/assets/7c2f8235-ee92-4117-9caf-9acd75547181)

IID_STUDY_URL=weasis://$dicom:rs --url "http://192.168.88.57:8080{{qidoBasePath}}" -r "studyUID={{studyUID}}" --query-ext "&includedefaults=false"

IID_URL_TARGET=_self

IID_PATIENT_URL=weasis://$dicom:rs --url "http://192.168.88.57:8080{{qidoBasePath}}" -r "patientID={{patientID}}" --query-ext "&includedefaults=false"

It was working two days ago

**Comment by ifaisalshahzad** [comment]
Recent changes in Chrome version 130.0.6723.59 and later, have affected how custom URL protocols like weasis:// are handled. This is causing issues with launching Weasis from DCM4CHEE. 
Chrome's new behavior:
- Custom protocols must now follow traditional URL rules: {CustomProtocol}://{website}:{port (numeric) optional}/{Path + string query optional}
- Multiple colons are no longer supported, except for specifying a port number.
- Spaces, double slashes (//), and additional colons in the path are no longer allowed.

Potential solutions:
Modify the Weasis launch URL to comply with Chrome's new rules, use URL encoding for special characters (e.g., replace spaces with %20)

as per [weasis article](https://weasis.org/en/getting-started/weasis-protocol/) To build an URI (weasis://path) without weasis-pacs-connector, you must choose one or more commands, encode the commands, and add the scheme weasis://

Problem Statement:
when encoding url as following and updating in DCM4CHEE
IID_STUDY_URL=weasis://$dicom%3Ars+--url+"https%3A//192.168.1.114:8080{{qidoBasePath}}"+-r+"studyUID={{studyUID}}"+--query-ext+"&includedefaults=false"
![image](https://github.com/user-attachments/assets/bc429c17-aa6a-4323-8c2d-2c44f3b09931)


after updating above, when navigating to view study, url getting encoded again as follows
weasis://$dicom%253Ars+--url+%22https%253A//192.168.1.114:8080/dcm4chee-arc/aets/DCM4CHEE/rs%22+-r+%22studyUID=1.3.6.1.4.1.5962.99.1.2786334768.1849416866.1385765836848.3.0%22+--query-ext+%22&includedefaults=false%22

how to stop DCM4CHEE encoding or handling the colon(:) encoding problem raised by chrome


Urgently waiting for community solution.

**Comment by shral** [comment]
As the dcm4che-arc-ui2 is only providing the given string as a URL, we cannot do much about how Weasis defined their custom web protocol, how they expect the URL to be, and how the browser (or in this case, Chrome) is handling the URL.
The only thing that I can think of for now and maybe would help from our side is to provide an additional parameter that would prevent the given string or URL from being encoded.

I added such a parameter (called `IID_ENCODE`) which can be set to false if you want that the dcm4che-arc-ui2 should not encode the URL (The parameter is of course optional, if is missing, it will encode as it did until now).

![image](https://github.com/user-attachments/assets/8f4bb4b7-1643-4b7a-9113-c2e4c6a921a8)

**Comment by nroduit** [comment]
I haven't done a real test, but looking at the code, I think this fix will only work partially because the injected parameters are not encoded when encodeCheck is false.

For example: `  patientID={{patientID}}`   where the value is equal to `  1168514^^^DCM4CHEE.023F5509.06CF2191` .
The result will render the URL invalid: `  patientID=1168514^^^DCM4CHEE.023F5509.06CF2191`  
It should be: `  patientID=1168514%5E%5E%5EDCM4CHEE.023F5509.06CF2191`  

So parameters (patientID, patientName and qidoBaseURL) must be encoded when encodeCheck is false, here is a proposal of patch:
```` diff
Index: dcm4chee-arc-ui2/src/app/study/study/study.component.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/dcm4chee-arc-ui2/src/app/study/study/study.component.ts b/dcm4chee-arc-ui2/src/app/study/study/study.component.ts
--- a/dcm4chee-arc-ui2/src/app/study/study/study.component.ts	(revision e32443dfe791e87790e1117258a313730f67124f)
+++ b/dcm4chee-arc-ui2/src/app/study/study/study.component.ts	(date 1732975120004)
@@ -1715,12 +1715,14 @@
             let patientName = _.get(model, "00100010.Value[0].Alphabetic");
             let dcmWebServicePath = this.studyWebService.selectedWebService.dcmWebServicePath;
             let qidoBaseURL = j4care.getUrlFromDcmWebApplication(this.studyWebService.selectedWebService,this.appService.baseUrl, true);
+
+            let encodeDynamicValue = (value) => encodeCheck ? (value || "") : encodeURIComponent(value || "");
             let replaceDoubleBraces = (url, result) => {
                 return url.replace(/{{(currentDateTime-)?(.+?)}}/g, (match, g1, g2) => {
                     if(g1){
                         return j4care.formatDate(j4care.createDateFromDuration(j4care.extractDurationFromValue(g2)),"yyyy-MM-dd");
                     }else{
-                        return result[g2] == null ? g2 : result[g2];
+                    return result[g2] == null ? g2 : encodeDynamicValue(result[g2]);
                     }
                 })
             };
````

There don't seem to be any solutions for passing the above-mentioned URI without encoding it directly in the configuration of the dcm4chee portal. 

Otherwise, the best solution is to use weasis-pacs-connector, which provides a URL. Soon there will be the component successor, weasis-manager, which in addition to providing URLs will use a new clientID to avoid passing the token in the URI (which can be truncated on Windows).

**Comment by PierrreWTZ** [comment]
Hello
I'm not a regular on your forum but I'd like to share my experience with you. When calling the weasis application with the command
IID_STUDY_URL=weasis://$dicom:rs --url ‘https://192.168.1.114:8080{{qidoBasePath}}’+-r+‘studyUID={{studyUID}}’+--query-ext+"&includedefaults=false’

I encourage you to simply add a ? after weasis:// to introduce this:

IID_STUDY_URL=weasis://?$dicom:rs --url ‘https://192.168.1.114:8080{{qidoBasePath}}’+-r+‘studyUID={{studyUID}}’+--query-ext+"&includedefaults=false’

Oh miracle everything will work normally again with all recent browsers

**Comment by nroduit** [comment]
Thanks @PierrreWTZ, indeed it seems that browsers have modified the encoding of URLs by differentiating the path from the query parameters.

As Weasis extracts the commands from the URI, it's not a problem to add '?' even for older versions. So I've updated the documentation for [integration](https://weasis.org/en/basics/customize/integration) and [weasis protocol](https://weasis.org/en/getting-started/weasis-protocol/).

**Comment by PierrreWTZ** [comment]
Hello,

 

If I look at your URL, I suggest you try this:

IID_STUDY_URL=weasis://?$dicom:rs --url "https://101.22.19:8080{{qidoBasePath}}" -r 'studyUID={{studyUID}}" --query-ext "&includedefaults=false" -H "Authorization: Bearer {{access_token}}"

IID_PATIENT_URL=weasis://?$dicom:rs --url "https://101.22.19:8080{{qidoBasePath}}" -r "patientID={{patientID}}" --query-ext "&includedefaults=false" -H "Authorization: Bearer {{access_token}}"

 

Best regards

Pierre

 

 

De : Revaninio ***@***.***> 
Envoyé : lundi, 24 mars 2025 12:25
À : dcm4che/dcm4chee-arc-light ***@***.***>
Cc : PierrreWTZ ***@***.***>; Mention ***@***.***>
Objet : Re: [dcm4che/dcm4chee-arc-light] Dicom files are not opening in weasis when click on view button of studies in dcm4chee ui (Issue #4592)

 

Hello guys, the behavior has indeed changed by adding the "?", Weasis is now lauching, but no image is displayed. It seems to still be an issue with the Token, as i use the secure version of DCM4CHEE ARC LIGHT. Any idea on how to fix it ?

Here are the URL i use :
IID_STUDY_URL=weasis://?$dicom:rs --url 'https://101.22.19:8080{{qidoBasePath}}'+-r+'studyUID={{studyUID}}'+--query-ext+ <https://101.22.19:8080%7B%7BqidoBasePath%7D%7D'+-r+'studyUID=%7B%7BstudyUID%7D%7D'+--query-ext+%22&includedefaults=false> "&includedefaults=false' -H 'Authorization: Bearer {{access_token}}'

IID_PATIENT_URL=weasis://?$dicom:rs --url 'https://101.22.19:8080{{qidoBasePath}}'+-r+'patientID={{patientID}} <https://101.22.19:8080%7B%7BqidoBasePath%7D%7D'+-r+'patientID=%7B%7BpatientID%7D%7D%22+--query-ext+%22&includedefaults=false> "+--query-ext+"&includedefaults=false' -H 'Authorization: Bearer {{access_token}}'

—
Reply to this email directly, view it on GitHub <https://github.com/dcm4che/dcm4chee-arc-light/issues/4592#issuecomment-2747791234> , or unsubscribe <https://github.com/notifications/unsubscribe-auth/AVASM4ZNZORR5ULORLFWSJD2V7TO7AVCNFSM6AAAAABQRLZZQGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONBXG44TCMRTGQ> .
You are receiving this because you were mentioned.  <https://github.com/notifications/beacon/AVASM4YLNKPT6AIMJCGZJ632V7TO7A5CNFSM6AAAAABQRLZZQGWGG33NNVSW45C7OR4XAZNMJFZXG5LFINXW23LFNZ2KUY3PNVWWK3TUL5UWJTVDY73YE.gif> Message ID: ***@***.*** ***@***.***> >

  <https://avatars.githubusercontent.com/u/89850146?s=20&v=4> Talos06100 left a comment (dcm4che/dcm4chee-arc-light#4592) <https://github.com/dcm4che/dcm4chee-arc-light/issues/4592#issuecomment-2747791234> 

Hello guys, the behavior has indeed changed by adding the "?", Weasis is now lauching, but no image is displayed. It seems to still be an issue with the Token, as i use the secure version of DCM4CHEE ARC LIGHT. Any idea on how to fix it ?

Here are the URL i use :
IID_STUDY_URL=weasis://?$dicom:rs --url 'https://101.22.19:8080{{qidoBasePath}}'+-r+'studyUID={{studyUID}}'+--query-ext+ <https://101.22.19:8080%7B%7BqidoBasePath%7D%7D'+-r+'studyUID=%7B%7BstudyUID%7D%7D'+--query-ext+%22&includedefaults=false> "&includedefaults=false' -H 'Authorization: Bearer {{access_token}}'

IID_PATIENT_URL=weasis://?$dicom:rs --url 'https://101.22.19:8080{{qidoBasePath}}'+-r+'patientID={{patientID}} <https://101.22.19:8080%7B%7BqidoBasePath%7D%7D'+-r+'patientID=%7B%7BpatientID%7D%7D%22+--query-ext+%22&includedefaults=false> "+--query-ext+"&includedefaults=false' -H 'Authorization: Bearer {{access_token}}'

**Reply by tctctc21212124-alt** [reply]
Did you find the solution for this? Weasis loads after clicking viewer but it has no dicom images.
